tests: Adds gcr.io Windows image test case

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind failing-test

/sig windows
/sig testing

**What this PR does / why we need it**:

Adds the test "should be able to pull image from gcr.io [NodeConformance]",
which will pull the the image "gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1".
The mentioned image is a manifest list, and it works for both
Windows Server 1803 and Windows Server 2019. The manifest list
will have to be amended when a new Windows Server is released.

Adds the test "should be able to pull from private registry with secret [NodeConformance]"
which will pull the image "gcr.io/authenticated-image-pulling/windows-nanoserver:v1".
The mentioned image is a manifest list, and it works for both
Windows Server 1803 and Windows Server 2019. The manifest list
will have to be amended when a new Windows Server is released.

The command passed to the Windows Container has been changed to
"ping -t localhost", which will keep the container in the Running state,
which is required and checked by the test.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #74134 
Fixes #74740

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

claudiubelu committed:
tests: Adds gcr.io Windows image test case

Adds the test "should be able to pull image from gcr.io [NodeConformance]",
which will pull the the image "gcr.io/kubernetes-e2e-test-images/windows-nanoserver:v1".
The mentioned image is a manifest list, and it works for both
Windows Server 1803 and Windows Server 2019. The manifest list
will have to be amended when a new Windows Server is released.

The command passed to the Windows Container has been changed to
"ping -t localhost", which will keep the container in the Running state,
which is required and checked by the test.

claudiubelu committed:
Removes [WindowsOnly] tag

Removes the [WindowsOnly] tag from the test:

"should be able to pull image from docker hub [WindowsOnly] [NodeConformance]"

neolit123 reviewed (commented):
@bclau 
thanks this LGTM.

but ideally we should replace the busy box image on L209 (1.5GB) with the same `windows-nanoserver:v1` uploaded to docker hub. because this way we are only going to download ~350 MB instead of 1.85GB

i have tested this locally.

/lgtm
/priority important-soon
/assign @timothysc

bclau commented:
> 
> but ideally we should replace the busy box image on L209 (1.5GB) with the same `windows-nanoserver:v1` uploaded to docker hub. because this way we are only going to download ~350 MB instead of 1.85GB

I would postpone that until the next cycle, given how close we are to the release. It is more of a wishlist item right now. Plus, as we've previously discussed, that image is based on the windows servercore image, which should already exist on the nodes, so the actual download is limited to just the busybox layers, which are a lot smaller.

neolit123 commented:
@bclau 

> Plus, as we've previously discussed, that image is based on the windows servercore image, which should already exist on the nodes, so the actual download is limited to just the busybox layers, which are a lot smaller.

should all images be based on nanoserver in the longterm?

claudiubelu committed:
tests: Adds private gcr.io Windows image test case

Adds the test "should be able to pull from private registry with secret [NodeConformance]"
which will pull the image "gcr.io/authenticated-image-pulling/windows-nanoserver:v1".
The mentioned image is a manifest list, and it works for both
Windows Server 1803 and Windows Server 2019. The manifest list
will have to be amended when a new Windows Server is released.

neolit123 commented:
/hold
"[WindowsOnly]" tag has to be removed.
pinged @bclau on slack after this was opened: 
https://github.com/kubernetes/kubernetes/issues/74740

bclau commented:
removed it from the new test.

neolit123 commented:
> removed it from the new test

has to be removed entirely:
https://github.com/kubernetes/kubernetes/blob/cb6d74ce70c731d719897e5fb7956ed8a6a98dbc/test/e2e/common/runtime.go#L360

neolit123 commented on test/e2e/common/runtime.go:
```diff
@@ -230,13 +230,18 @@ while true; do sleep 1; done
-			imagePullTest := func(image string, hasSecret bool, expectedPhase v1.PodPhase, expectedPullStatus bool) {
+			imagePullTest := func(image string, hasSecret bool, expectedPhase v1.PodPhase, expectedPullStatus bool, windowsImage bool) {
+				command := []string{"/bin/sh", "-c", "while true; do sleep 1; done"}
+				if windowsImage {
```
isn't this the same as `framework.NodeOSDistroIs("windows")`, which avoids passing the flag?

bgrant0607 replied:
Given that the specified image needs to be aligned with the command, I like windowsImage as a parameter. I could also imagine an enum, for call site readability, but that would introduce the need to check for other values, so I don't really recommend it.

bgrant0607 commented:
cc @dims

neolit123 commented:
/lgtm

neolit123 commented:
@bclau the auth windows image is up too.
https://github.com/kubernetes/kubernetes/issues/74134#issuecomment-468471758

so you might as well add it as part of this PR.

also please remove the release note:
> Added a Windows equivalent test to ``should

`NONE` is sufficient here i'd say.

neolit123 commented on test/e2e/common/runtime.go:
```diff
@@ -328,39 +333,50 @@ while true; do sleep 1; done
+				imagePullTest(image, false, v1.PodRunning, false, true)
 			})
 
 			It("should not be able to pull from private registry without secret [NodeConformance]", func() {
```
aren't we supposed to test this on Windows too?

neolit123 reviewed (commented):
/lgtm
/assign @timothysc

claudiubelu committed:
tests: Adds TODO for Windows only test

It has been suggested to replace the "e2eteam/busybox:1.29" image
used in the test "should be able to pull image from docker hub [NodeConformance]"
with a nanoserver image manifest list.

Adds a TODO for it.

PatrickLang commented:
/milestone v1.14

neolit123 commented:
needs approval at this point.

dims commented:
/hold cancel
/lgtm

timothysc commented on test/e2e/common/runtime.go:
```diff
@@ -328,39 +333,50 @@ while true; do sleep 1; done
 				framework.SkipUnlessNodeOSDistroIs("windows")
 				image := "e2eteam/busybox:1.29"
-				imagePullTest(image, false, v1.PodRunning, false)
+				imagePullTest(image, false, v1.PodRunning, false, true)
```
FWIW I get why this is changed to true, but busybox/windows test?  This specific test is a little weird and needs better description IMO.  I don't understand why this is windows specific, or why the logic exists above.

neolit123 replied:
the busybox here that is pulled will end up being Windows based (manifest list), though a comment is not present about this in the code.

@bclau might as well use an official Microsoft image here?

neolit123 replied:
why not push the same `v1` image we now have on GCR to the docker hub account....?

bclau replied:
Indeed, the pulled image is manifest list with only Windows images. This is the Windows equivalent test of "should be able to pull image from docker hub [LinuxOnly] [NodeConformance]"

dims replied:
@PatrickLang let's wait for this conversation to be resolved before we wrap up this PR

michmike replied:
@neolit123  I understand that using the official msft image is better and we can moving forward but we are using busybox in many tests so far. we don't want to make all these changes now 2 days before code freeze. are you ok with that? we can file an issue and fix for 1.15

neolit123 replied:
@michmike i don't want to block anything.
@timothysc questioned the usage of busybox, so i proposed solutions.

also i'm not an approver for `/test` so the final call is not mine.

michmike replied:
i agree with your proposal, i was just asking to postpone it a little bit :)

timothysc reviewed (commented):
Need quick feedback before approve.

timothysc commented:
/approve 
/lgtm cancel 
/hold  

Please add the TODO, open an issue for tracking (/assign @stevesloka) and we can call it good enough.

timothysc reviewed (approved):
/lgtm 
/hold cancel

bclau commented:
/test pull-kubernetes-integration

PatrickLang commented:
/retest

PatrickLang commented:
/retest pull-kubernetes-kubemark-e2e-gce-big
